opm/simulators/linalg/bda/openclSolverBackend.cpp

```diff
@@ -251,13 +276,19 @@ void openclSolverBackend<block_size>::gpu_pbicgstab(WellContributions& wellContr
         t_rest.stop();
 
         // pw = prec(p)
-        t_prec.start();
-        prec->apply(d_p, d_pw);
-        t_prec.stop();
+        t_bilu0.start();
+        bilu0->apply(d_p, d_pw);
+        t_bilu0.stop();
+
+        if (use_cpr) {
+            t_cpr.start();
+            cpr->apply(d_p, d_pw);
+            t_cpr.stop();
+        }
```

Comment:
This looks strange to me. When I read CPR, I assume that it it the full preconditioner (iilu on the fine system, plus the correction on the pressure only system). Here it seems like the CPR is just rerstriction to pressure system, solve for correction, prolongate and add correction.

What is the reason, why we can't put the whole CPR in a class/struct and only have one application of the preconditioner here that does it all? (Why can either use virtual inheritance here, or type erasure with std::function)

Thus we would get rid of all the rather ugly if clauses introduced here.

Reply:
I combined them, now `openclSolver` has a `cpr` object, which has a `bilu0` object.
I moved the `use_cpr` bool to the `CPR` class, and renamed it to `use_amg`, but I'm not sure if that is the best name. Perhaps `pressure_correction` is a better name.